### ISSUE TITLE
Project status update?

### ISSUE THREAD

**Issue by pejaham** [salat/salat#160]
It has been quiet for a while here - does anybody have news on the status of the project?

I updated a current project to use Casbah 2.8.2/MongoDB 3.0.8 as suggested here in the forum and it works fine so far.

Since I will start a project in the near future I would like to plan ahead if I should stick with salat or start using ReactiveMongo - but I would prefer salat with Casbah 3/Mongo Java Driver 3 over ReactiveMongo. I would not mind waiting for an asynchronous API - if the library will updated at all, the asynchronous API being implemented later.

Thanks and regards,
Peter

**Comment by sief** [comment]
+1

**Comment by noahlz** [comment]
Hello all. As you may be aware, @rktoomey is the primary owner of this project. While I'm around for now to answer questions and triage issues (like this), I can't speak as to upcoming plans. I've reached out to her and hopefully we'll hear something soon.

The reason it's been quiet here is the usual mundane reasons things slow down on volunteer, OSS projects like this - work and major life events sometimes reduce the time we have for development work, which has to happen on increasingly sparse free time on nights and weekends.

Thanks for your support of the project, and let me know if there's anything I can do to help in the short term!

**Comment by sief** [comment]
Thanks for your support, Noah, appreciate it. 

My biggest problem right now is the casbah version. Is there any specific reason that  https://github.com/salat/salat/pull/157 hasn't been merged yet?

thanks again
Simon

**Comment by noahlz** [comment]
Hello - sincere apologies for painfully belated response.

The reason that #157 hasn't been merged is because @rktoomey is the lead maintainer of Salat, and while I'm supporting the project by responding to issues (like this) and perhaps resolving small issues in the near future - she's the one who will review and merge large PRs like this.

If you need to get unblocked, I recommend forking the project and proceeding with the PR for #157.  That's actually what we are currently doing here at Novus Partners - we have an internally forked version of Salat that fixes 1 or 2 bugs that I have not had the time to formally release to the outside world.

**Comment by noahlz** [comment]
Greetings all. At long last, I finally have carved out some time to work on this, and have an update.

Our current plan for salat is to have three branches: 

| Branch          | Latest Version | Description |
|-----------------|----------------|-------------|
| `master`        | None           | Home of `2.0.0-SNAPSHOT`. @rktoomey owns this branch. Future development of salat 2.0 is on this branch, and up to her. |
| `1.9.x-branch`  | `1.9.9` (`1.9.10` coming soon)       |  Home of support for the 1.9.x releases. This branch will be for legacy support, with a conservative policy on bug fixes / enhancements. Don't expect a lot of activity on this branch. |
| `1.10.x-branch` | (Forthcoming)           | Home of `1.10.x` versions, and all community-submitted but fixes / enhancements. Will be branched off `1.9.10` (when that is released) |

So, `1.10.x-branch` and `1.10.x` releases will be forthcoming, and many of the major outstanding PRs and such will go against that branch. For example, long-awaited Mongo and Casbah upgrades.  This will be a "community" branch, meaning Salat users will submit the enhancements - and bug fixes. Releases will be much more frequent, driven by users, and might have corresponding volatility in code quality.

**tl;dr: many outstanding PRs will need to be re-pointed to the forthcoming `1.10.x-branch` branch, when it is available.**

Once I've released `1.9.10` and pushed the `1.9.x-branch` here, I will push a `1.10.x-branch`. Then we can start updating / revising PRs that were against `master` to be against `1.10.x-branch` instead - and start merging / cutting `1.10.x` releases.

The `1.9.x-branch` branch is arriving soon (if not by the time you are reading this). I spent most of today working on `1.9.10-SNAPSHOT` and getting it ready for release.  

Anyway, please let us know if this seems reasonable and leave your feedback. Thanks!

**Comment by sief** [comment]
@noahlz this looks very reasonable, really appreciate it, thanks!

**Comment by noahlz** [comment]
One other thought is package names.  Since 1.9.x is for legacy support, it will stay in package `com.novus.salat` but 1.10.x and 2.x.x will change the package to just `salat` - so `salat.grater` `salat.globals` etc.

The reasoning being that Novus Partners is (if not obvious) no longer officially sponsoring Salat (although we still use it internally), and the package name should reflect that.

Also, there will probably be breaking changes between 1.9.x and later versions.

Any thoughts on this?  Is it better to keep the package name the same, to avoid namespace / import pollution?

**Comment by sief** [comment]
maybe it would make sense to keep the package names in sync with the buildOrganization, so in this case it would be `com.github.salat`. This follows the reverse domain pattern and also avoids potential namespace conflicts. Just an idea....

**Comment by noahlz** [comment]
The issue I have with that is that Salat is not a project of Github (the company), it is just hosted here. There is a rather established precedent of having projects hosted on github, but not using the `com.github` - for example `ScalaZ`

The only reason we changed our project coordinates to `com.github` was to continue hosting the artifacts in the Sonatype OSS repo last year.

Personally, I would prefer to change the package names to just `salat.*`

**Comment by sief** [comment]
you're right, the projects `cats` and `shapeless` don't use the reverse domain pattern either. I agree that there's no need for this domain prefix.

**Comment by noahlz** [comment]
Salat `1.9.10` is officially released and published to Sonatype OSS. Example project: 
https://github.com/noahlz/salat-examples

RELATED: I now am the proud owner of one of those fancy PGP keys.

This coming weekend (July 16 - 17, 2016) I will be setting up the `1.10.x-branch` and we can start planning what outstanding PRs to merge in, Mongo/Casbah upgrade being no doubt the top priority. 

Probably renaming the packages to `salat.*` and moving around source files will be the last step of the `1.10.x-branch`, because I don't want to create a lot of work for authors of existing PRs.

Once `1.10.x-branch` is pushed, I'm likely to close this ticket, unless anybody has further questions.

**Comment by noahlz** [comment]
Apologies for the delay - `1.10.0-SNAPSHOT` is now available, with #145 as its first change - Casbah 2.8

Any other nominations on what PRs should go into the 1.10.x branch?  

My thought is that 1.10.x-branch will be the "default" branch of this project going forward. As mentioned before, `master` will be reserved for major rewrites, completely incompatible changes, etc.

**Comment by noahlz** [comment]
Note that the SNAPSHOT release is only available from the Sonatype OSS repository - it doesn't get promoted to Central.

**Comment by noahlz** [comment]
#158 Casbah 2.8.2 / Mongo 3.x support is available for testing in the lastest `1.10.0-SNAPSHOT` now available in the Sonatype OSS snapshot repo.

Feedback appreciated!

**Comment by noahlz** [comment]
Hello all - a quick update - I decided that I want to knock out a few more PRs and issues for `1.10.0` - see the latest activity on `1.10.x-branch` and PRs targeting that branch.

If you have any requests or PRs to issue - please send them along!

I hope to publish a candidate snapshot next weekend (Aug 27 - 28).

**Comment by noahlz** [comment]
Just published latest `1.10.0-SNAPSHOT`

Here are the notes thus far (as of 53a7e28)

## Dependency Upgrades
- #145 Upgrade to Casbah 2.8.2 (@danslapman, @sief, @noahlz)

## Features
- #115 Add `findAndModify` to BaseDAOMethods (@noahlz)
- #143 Log more error details in case of deserialization problem (@noahlz, @slorber)
- #155 Improve support for binary data such as `Seq[Byte]` in case classes. (@glorat)
- #171 Support `Map[String, Any]`, `List[Any], Option[Any]` (@dieu, @noahlz)
- #173 Improve error message for `Map[String, List[_]` de-serialization error. (@noahlz)

# Compatibility
- If you create a `SalatDAO` using a collection created from a `MongoClient`, its methods may throw a `SalatDAOError` having a `MongoException` root cause. `SalatDAOError` now extends `RuntimeException` instead of `java.lang.Error`, and the constructor signature has changed accordingly.
- A `SalatDAO` created using a `MongoCollection` from a `MongoConnection` will attempt attempt to detect errors from the `WriteResult`, but is not guaranteed (from the Mongo 3.x upgrade notes: "[getLastError] does not work reliably in the 2.x series and there is no way to make it work reliably" - http://mongodb.github.io/mongo-java-driver/3.0/whats-new/upgrading/). Be sure to use `MongoClient`.
- Introduced a new subclass of `Error`, `SalatGlitch`, that specifically identifies errors that occur during serialization / deserialization. Note that there are still many points in code that throw RuntimeException (via sys.error) - overhaul/normalization of Salat's errors remains an outstanding TODO.

**Comment by noahlz** [comment]
This is the last pending issue before `1.10` is released: #178 (resolves #148) 

I need some feedback on this one if anybody is feeling generous. It will certainly break existing code (cause errors where previously code had failed silently on bad data)

**Comment by noahlz** [comment]
Just published a new `1.10.0-SNAPSHOT` that includes #148 / #178 

I also added a long comment to #178 that explains my thinking on this PR, if anyone's interested.

Last step is #177 "Release Salat 1.10" which will be the final steps: drop `com.novus` from the package, update release notes and README, etc. I'm aiming to get this done this week with the final release next Sunday - if possible.

If anybody is looking to contribute, I've added a project label "Help Needed" and labeled some issues and PRs appropriately.

**Comment by noahlz** [comment]
PR for 1.10.0 release: #180

**Comment by noahlz** [comment]
Published `1.10.0-SNAPSHOT` which changes `com.novus.salat._` to `salat._`

**Comment by noahlz** [comment]
I have published Salat version 1.10.0 which has support for Casbah 2.8.2 and Mongo 3.x

Release notes: https://github.com/salat/salat/blob/1.10.x-branch/notes/1.10.0.markdown

**Comment by slorber** [comment]
thanks for your work on this @noahlz it is valuable even if there's not a lot of congratulations atm it will be useful to many of us, now or in the futur :)
